### DISCUSSION TITLE
Jtl Reporter integration with Terraform

### DISCUSSION THREAD

**Discussion by jmodeorain14** [ludeknovy/jtl-reporter#225] [Q&A]
Hi,

As part of my PoC project, I am using Terraform to provision infrastructure and to execute a JMeter performance test.

I was wondering if there is a way to implement the start and stop of the Jtl Reporter docker container either locally/on an EC2 instance so that it is not always running 24/7.

Thanks in advance!

**Comment by ludeknovy** [comment]
Hi @jmodeorain14 ! I guess there must be a way, but I am not familiar with terraform myself. Would you be willing to put it together and share it with the community?

**Comment by jmodeorain14** [comment]
Thanks for getting back to me @ludeknovy!

I'm trying to work out if this is even possible. If I get something working, I'll let you know.

**Reply by ludeknovy** [reply]
Thanks!

Although, I think that the initial idea of the app was to be always on, as the flow of analysing the results could be asynchronous.
But if turning it off fits your workflow, why not :)

**Reply by jmodeorain14** [reply]
I have no problem running the docker container locally.

I'm creating the EC2 instances that are running the test via Terraform, and currently I'm trying to get the JMeter Master EC2 instance to send the test results (.jtl) file to a locally running docker container with Jtl Reporter using a curl request as stated [here](https://jtlreporter.site/docs/guides/api-data-upload).

My concern here is that if I shut down my laptop, the container will stop and all of the data on the container (test results) will be lost.

Also, if I run the docker container in an AWS EC2 instance, I will be paying for it as long as it runs.

I'm trying to figure out the best way to proceed here.

**Reply by ludeknovy** [reply]
Yes, I got that :)

> Also, if I run the docker container in an AWS EC2 instance, I will be paying for it as long as it runs.

Yes, it's like that. How much would it be? What about running somewhere else cheaper? Also, you will need to have kind of logic to not to turn it off during processing, so you will need to wait until a report is fully processed.

What I was trying to say is that, to me (in the context(s) I used to work), this does not make that much sense to turn it off, because if there is more people using it, you never know when it's needed. Especially if you use features like notifications, anonymous report sharing, you never know when it will be accessed. In such a case, maybe a static HTML report would be better?

**Comment by jmodeorain14** [comment]
Hi again @ludeknovy.
I've gone down the EC2 instance route for now.
I have the Jtl Reporter running on a persistent EC2 instance.
But when I am trying to run the curl request to post the test results from another EC2 instance to the Jtl Reporter instance, it doesn't work.
I've even tried running the curl request manually from the EC2 instance where the test runs, but with no success.
Here's an example of the curl request I am trying to use:

curl -X POST 'http://{Jtl_Reporter_PublicIPv4}:2020/api/projects/jmeterterraformproject/scenarios/jmeterterraformscenario/items' \
  -H 'x-access-token: 123456' \
  -F 'kpi=@/home/ubuntu/test.jtl' \
  -F 'environment=testenv'
  
  Am I doing something wrong here?

**Reply by ludeknovy** [reply]
Hi @jmodeorain14 .
What do you mean by `it does not work`?  Do you get any response from the server? Do you get any error message?

**Reply by jmodeorain14** [reply]
Here is the actual curl request I am trying to send:

curl -v -X POST 'http://{Jtl_Reporter_Public_IPv4}:5000/api/projects/jmeterterraformproject/scenarios/jmeterterraformscenario/items' \
      -H 'x-access-token: {API_token}' \
      -F "kpi=@\"/home/ubuntu/apache-jmeter-5.5/bin/TestResults/${filename}\"" \
      -F 'environment=testenv'


When running the command from within a bash script, there are no obvious errors returned.

When trying to run the same request manually, I see the following response which indicates that it was successful, however, the data does not show up in the Jtl Reporter:

Note: Unnecessary use of -X or --request, POST is already inferred.
*   Trying {Jtl_Reporter_Public_IPv4}:5000...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
^M  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0* Connected to {Jtl_Reporter_Public_IPv4} ({Jtl_Reporter_Public_IPv4}) port 5000 (#0)
> POST /api/projects/jmeterterraformproject/scenarios/jmeterterraformscenario/items HTTP/1.1^M
> Host: {Jtl_Reporter_Public_IPv4}:5000^M
> User-Agent: curl/7.81.0^M
> Accept: */*^M
> x-access-token: {API_token}^M
> Content-Length: 665^M
> Content-Type: multipart/form-data; boundary=------------------------ee9265578cbe2bf1^M
> ^M
} [665 bytes data]
* We are completely uploaded and fine
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK^M
< Content-Security-Policy: default-src 'self';base-uri 'self';font-src 'self' https: data:;form-action 'self';frame-ancestors 'self';img-src 'self' data:;object-src 'none';script-src 'self';script-src-attr 'none';style-src 'self' https: 'unsafe-inline';upgrade-insecure-requests^M< Cross-Origin-Embedder-Policy: require-corp^M
< Cross-Origin-Opener-Policy: same-origin^M
< Cross-Origin-Resource-Policy: same-origin^M
< Origin-Agent-Cluster: ?1^M
< Referrer-Policy: no-referrer^M
< Strict-Transport-Security: max-age=15552000; includeSubDomains^M
< X-Content-Type-Options: nosniff^M
< X-DNS-Prefetch-Control: off^M
< X-Download-Options: noopen^M
< X-Frame-Options: SAMEORIGIN^M
< X-Permitted-Cross-Domain-Policies: none^M
< X-XSS-Protection: 0^M
< Access-Control-Allow-Origin: *^M
< Access-Control-Allow-Methods: *^M
< Access-Control-Allow-Headers: Origin, X-Requested-With, x-access-token, Content-Type, Accept^M
< Content-Type: application/json; charset=utf-8^M
< Content-Length: 49^M
< ETag: W/"31-+phNSlwOBDlClZ0tvJ5/oxUTPBc"^M
< Vary: Accept-Encoding^M
< Date: Thu, 03 Aug 2023 07:09:58 GMT^M
< Connection: keep-alive^M
< Keep-Alive: timeout=5^M
< ^M
{ [49 bytes data]
^M100   714  100    49  100   665   1591  21601 --:--:-- --:--:-- --:--:-- 23800
* Connection #0 to host {Jtl_Reporter_Public_IPv4} left intact

**Reply by ludeknovy** [reply]
> I see the following response which indicates that it was successful

if you are sure the data was sent correctly to the server, you might want to check the logs for any errors -> `docker logs -f <be-service-docker-id>.`

**Reply by ludeknovy** [reply]
@jmodeorain14 Did you find any error(s) in the logs?

**Reply by jmodeorain14** [reply]
Thanks for your support @ludeknovy.

I just had a chance to try this again.

Here is the curl request I am sending (the name of the .jtl file is unique each time, hence use of the "filename" variable):

curl -v -X POST 'http://3.253.61.230:5000/api/projects/jmeterterraformproject/scenarios/jmeterterraformscenario/items' \
  -H 'x-access-token: at-628a3d48-cac6-4854-a910-38f90c93011a' \
  -F "kpi=@"/home/ubuntu/apache-jmeter-5.5/bin/TestResults/${filename}"" \
  -F 'environment="testenv"'

Here is the response I see for the request on the machine where the request is run (note an "itemId" is returned):

Post the test result file to the Jtl Reporter
Note: Unnecessary use of -X or --request, POST is already inferred.
Trying {Private_IPv4_Address}:5000...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
^M  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0* Connected to {Private_IPv4_Address} ({Private_IPv4_Address}) port 5000 (#0)
> POST /api/projects/jmeterterraformproject/scenarios/jmeterterraformscenario/items HTTP/1.1^M
> Host: {Private_IPv4_Address}:5000^M
> User-Agent: curl/7.81.0^M
> Accept: */*^M
> x-access-token: {API_token}^M
> Content-Length: 80516^M
> Content-Type: multipart/form-data; boundary=------------------------a64cf8c8a6bac5ec^M
> ^M
} [65536 bytes data]
We are completely uploaded and fine
Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK^M
< Content-Security-Policy: default-src 'self';base-uri 'self';font-src 'self' https: data:;form-action 'self';frame-ancestors 'self';img-src 'self' data:;object-src 'none';script-src 'self';script-src-attr 'none';style-src 'self' https: 'unsafe-inline';upgrade-insecure-requests^M
< Cross-Origin-Embedder-Policy: require-corp^M
< Cross-Origin-Opener-Policy: same-origin^M
< Cross-Origin-Resource-Policy: same-origin^M
< Origin-Agent-Cluster: ?1^M
< Referrer-Policy: no-referrer^M
< Strict-Transport-Security: max-age=15552000; includeSubDomains^M
< X-Content-Type-Options: nosniff^M
< X-DNS-Prefetch-Control: off^M
< X-Download-Options: noopen^M
< X-Frame-Options: SAMEORIGIN^M
< X-Permitted-Cross-Domain-Policies: none^M
< X-XSS-Protection: 0^M
< Access-Control-Allow-Origin: *^M
< Access-Control-Allow-Methods: *^M
< Access-Control-Allow-Headers: Origin, X-Requested-With, x-access-token, Content-Type, Accept^M
< Content-Type: application/json; charset=utf-8^M
< Content-Length: 49^M
< ETag: W/"31-miMr/gl9xWoUjU66epJdIEyn5Ng"^M
< Vary: Accept-Encoding^M
< Date: Fri, 04 Aug 2023 09:36:36 GMT^M
< Connection: keep-alive^M
< Keep-Alive: timeout=5^M
< ^M
{ [49 bytes data]
^M100 80565  100    49  100 80516   1547  2483k --:--:-- --:--:-- --:--:-- 2537k
Connection #0 to host {Private_IPv4_Address} left intact
{"itemId":"c03754b6-26b2-43a2-9ad2-4c95a41faad5"}

Again, I don't see any information about the test showing up in the Jtl Reporter.

Here is what I see in the docker logs forr the jtl-reporter-be docker container:

{"level":"info","message":"User 15b2f921-c555-435b-9ffc-e23ff2d0b376 with role admin accessing a resource with allowed roles: operator,admin"}
{"level":"info","message":"Starting new item processing for scenario: jmeterterraformscenario"}
{"level":"info","message":"POST /api/projects/jmeterterraformproject/scenarios/jmeterterraformscenario/items 200 10ms","meta":{}}
{"level":"info","message":"Starting KPI file streaming and saving to db, item_id: f27f868d-4bf1-452b-bc5d-3852b6ddab32"}
{"level":"error","message":"Error while parsing data: not a number"}
{"level":"error","message":"Error while parsing data: not a number"}
{"level":"error","message":"Error while parsing data: not a number"}
{"level":"error","message":"Error while parsing data: not a number"}
{"level":"error","message":"Error while parsing data: not a number"}
(+ 495 more identical rows)
{"level":"error","message":"Error while processing item: f27f868d-4bf1-452b-bc5d-3852b6ddab32: TypeError: Cannot generate an INSERT from an empty array."}
{"level":"info","message":"User 15b2f921-c555-435b-9ffc-e23ff2d0b376 with role admin accessing a resource with allowed roles: readonly,operator,admin"}
{"level":"info","message":"GET /api/projects/jmeterterraformproject/scenarios/jmeterterraformscenario/processing-items?environment= 200 5ms","meta":{}}
{"level":"info","message":"User 15b2f921-c555-435b-9ffc-e23ff2d0b376 with role admin accessing a resource with allowed roles: readonly,operator,admin"}
{"level":"info","message":"GET /api/projects/jmeterterraformproject/scenarios/jmeterterraformscenario/processing-items?environment= 304 4ms","meta":{}}
{"level":"info","message":"User 15b2f921-c555-435b-9ffc-e23ff2d0b376 with role admin accessing a resource with allowed roles: readonly,operator,admin"}
{"level":"info","message":"GET /api/projects/jmeterterraformproject/scenarios/jmeterterraformscenario/processing-items?environment= 304 5ms","meta":{}}
{"level":"info","message":"User 15b2f921-c555-435b-9ffc-e23ff2d0b376 with role admin accessing a resource with allowed roles: readonly,operator,admin"}
{"level":"info","message":"GET /api/projects/jmeterterraformproject/scenarios/jmeterterraformscenario/processing-items?environment= 304 5ms","meta":{}}
{"level":"info","message":"User 15b2f921-c555-435b-9ffc-e23ff2d0b376 with role admin accessing a resource with allowed roles: readonly,operator,admin"}
{"level":"info","message":"GET /api/projects 200 6ms","meta":{}}
{"level":"info","message":"User 15b2f921-c555-435b-9ffc-e23ff2d0b376 with role admin accessing a resource with allowed roles: admin"}
{"level":"info","message":"GET /api/users 304 9ms","meta":{}}
{"level":"info","message":"User 15b2f921-c555-435b-9ffc-e23ff2d0b376 with role admin accessing a resource with allowed roles: readonly,operator,admin"}
{"level":"info","message":"User 15b2f921-c555-435b-9ffc-e23ff2d0b376 with role admin accessing a resource with allowed roles: readonly,operator,admin"}
{"level":"info","message":"GET /api/projects/jmeterterraformproject/scenarios/jmeterterraformscenario/processing-items?environment= 304 15ms","meta":{}}
{"level":"info","message":"GET /api/projects/jmeterterraformproject/scenarios/jmeterterraformscenario/environment 304 28ms","meta":{}}
{"level":"info","message":"User 15b2f921-c555-435b-9ffc-e23ff2d0b376 with role admin accessing a resource with allowed roles: readonly,operator,admin"}
{"level":"info","message":"User 15b2f921-c555-435b-9ffc-e23ff2d0b376 with role admin accessing a resource with allowed roles: readonly,operator,admin"}
{"level":"info","message":"GET /api/projects/jmeterterraformproject/scenarios/jmeterterraformscenario/items?limit=15&offset=0&environment= 304 48ms","meta":{}}
{"level":"info","message":"User 15b2f921-c555-435b-9ffc-e23ff2d0b376 with role admin accessing a resource with allowed roles: readonly,operator,admin"}
{"level":"info","message":"GET /api/projects/jmeterterraformproject/scenarios/jmeterterraformscenario 304 9ms","meta":{}}
{"level":"info","message":"User 15b2f921-c555-435b-9ffc-e23ff2d0b376 with role admin accessing a resource with allowed roles: readonly,operator,admin"}
{"level":"info","message":"GET /api/projects/jmeterterraformproject/scenarios/jmeterterraformscenario/notifications 304 5ms","meta":{}}
{"level":"info","message":"GET /api/projects/jmeterterraformproject/scenarios/jmeterterraformscenario/trends?environment= 304 58ms","meta":{}}
{"level":"info","message":"User 15b2f921-c555-435b-9ffc-e23ff2d0b376 with role admin accessing a resource with allowed roles: readonly,operator,admin"}
{"level":"info","message":"GET /api/projects/jmeterterraformproject/scenarios/jmeterterraformscenario 304 7ms","meta":{}}
{"level":"info","message":"User 15b2f921-c555-435b-9ffc-e23ff2d0b376 with role admin accessing a resource with allowed roles: readonly,operator,admin"}
{"level":"info","message":"GET /api/projects 304 7ms","meta":{}}
{"level":"info","message":"User 15b2f921-c555-435b-9ffc-e23ff2d0b376 with role admin accessing a resource with allowed roles: admin"}
{"level":"info","message":"User 15b2f921-c555-435b-9ffc-e23ff2d0b376 with role admin accessing a resource with allowed roles: readonly,operator,admin"}
{"level":"info","message":"User 15b2f921-c555-435b-9ffc-e23ff2d0b376 with role admin accessing a resource with allowed roles: readonly,operator,admin"}
{"level":"info","message":"User 15b2f921-c555-435b-9ffc-e23ff2d0b376 with role admin accessing a resource with allowed roles: readonly,operator,admin"}
{"level":"info","message":"GET /api/projects/jmeterterraformproject/scenarios/jmeterterraformscenario/environment 304 9ms","meta":{}}
{"level":"info","message":"GET /api/projects/jmeterterraformproject/scenarios/jmeterterraformscenario/items?limit=15&offset=0&environment= 304 15ms","meta":{}}
{"level":"info","message":"GET /api/users 304 19ms","meta":{}}
{"level":"info","message":"GET /api/projects/jmeterterraformproject/scenarios/jmeterterraformscenario/trends?environment= 304 22ms","meta":{}}
{"level":"info","message":"User 15b2f921-c555-435b-9ffc-e23ff2d0b376 with role admin accessing a resource with allowed roles: readonly,operator,admin"}
{"level":"info","message":"GET /api/projects/jmeterterraformproject/scenarios/jmeterterraformscenario/processing-items?environment= 304 25ms","meta":{}}
{"level":"info","message":"User 15b2f921-c555-435b-9ffc-e23ff2d0b376 with role admin accessing a resource with allowed roles: readonly,operator,admin"}
{"level":"info","message":"GET /api/projects/jmeterterraformproject/scenarios/jmeterterraformscenario 304 7ms","meta":{}}
{"level":"info","message":"User 15b2f921-c555-435b-9ffc-e23ff2d0b376 with role admin accessing a resource with allowed roles: readonly,operator,admin"}
{"level":"info","message":"GET /api/projects/jmeterterraformproject/scenarios/jmeterterraformscenario/notifications 304 6ms","meta":{}}
{"level":"info","message":"User 15b2f921-c555-435b-9ffc-e23ff2d0b376 with role admin accessing a resource with allowed roles: readonly,operator,admin"}
{"level":"info","message":"GET /api/projects/jmeterterraformproject/scenarios/jmeterterraformscenario 304 7ms","meta":{}}
{"level":"info","message":"User 15b2f921-c555-435b-9ffc-e23ff2d0b376 with role admin accessing a resource with allowed roles: readonly,operator,admin"}
{"level":"info","message":"GET /api/projects 304 9ms","meta":{}}
{"level":"info","message":"User 15b2f921-c555-435b-9ffc-e23ff2d0b376 with role admin accessing a resource with allowed roles: readonly,operator,admin"}
{"level":"info","message":"GET /api/projects/jmeterterraformproject/scenarios 304 7ms","meta":{}}
{"level":"info","message":"User 15b2f921-c555-435b-9ffc-e23ff2d0b376 with role admin accessing a resource with allowed roles: readonly,operator,admin"}
{"level":"info","message":"GET /api/projects 304 5ms","meta":{}}
{"level":"info","message":"User 15b2f921-c555-435b-9ffc-e23ff2d0b376 with role admin accessing a resource with allowed roles: admin"}
{"level":"info","message":"User 15b2f921-c555-435b-9ffc-e23ff2d0b376 with role admin accessing a resource with allowed roles: readonly,operator,admin"}
{"level":"info","message":"GET /api/users 304 7ms","meta":{}}
{"level":"info","message":"GET /api/projects/jmeterterraformproject/scenarios 304 8ms","meta":{}}
{"level":"info","message":"User 15b2f921-c555-435b-9ffc-e23ff2d0b376 with role admin accessing a resource with allowed roles: readonly,operator,admin"}
{"level":"info","message":"GET /api/projects 304 4ms","meta":{}}
{"level":"info","message":"User 15b2f921-c555-435b-9ffc-e23ff2d0b376 with role admin accessing a resource with allowed roles: readonly,operator,admin"}
{"level":"info","message":"User 15b2f921-c555-435b-9ffc-e23ff2d0b376 with role admin accessing a resource with allowed roles: admin"}
{"level":"info","message":"GET /api/users 304 4ms","meta":{}}
{"level":"info","message":"GET /api/projects/jmeterterraformproject/scenarios 304 8ms","meta":{}}

**Reply by jmodeorain14** [reply]
Note all the errors stating "{"level":"error","message":"Error while parsing data: not a number"}".

Here is a small sample of the contents of the .jtl file I am trying to upload:

timeStamp;elapsed;label;responseCode;responseMessage;threadName;success;bytes;sentBytes;grpThreads;allThreads;URL;Latency;SampleCount;ErrorCount;Hostname;IdleTime;Connect
1691141758840;254;POC01_NH_01_NavigateToHomepage;200;OK;10.11.1.209-Navigate To Homepage 1-1;true;251556;306;1;1;https://www.{website}.com/;187;1;0;ip-10-11-1-209;0;168
1691141759920;82;POC01_NH_01_NavigateToHomepage;200;OK;10.11.1.209-Navigate To Homepage 1-1;true;251553;306;1;1;https://www.{website}.com/;47;1;0;ip-10-11-1-209;0;35
1691141760007;84;POC01_NH_01_NavigateToHomepage;200;OK;10.11.1.209-Navigate To Homepage 1-1;true;251561;306;1;1;https://www.{website}.com/;47;1;0;ip-10-11-1-209;0;35
1691141760094;79;POC01_NH_01_NavigateToHomepage;200;OK;10.11.1.209-Navigate To Homepage 1-1;true;251561;306;1;1;https://www.{website}.com/;47;1;0;ip-10-11-1-209;0;33
1691141760178;101;POC01_NH_01_NavigateToHomepage;200;OK;10.11.1.209-Navigate To Homepage 1-1;true;251561;306;1;1;https://www.{website}.com/;61;1;0;ip-10-11-1-209;0;34

**Reply by ludeknovy** [reply]
Ahhh, now I see it. The .jtl file contains unsupported separator `;` if you replace it with a comma `,` it will work.

**Reply by jmodeorain14** [reply]
Let me try that quickly and I'll let you know how it goes.

**Reply by jmodeorain14** [reply]
It's working now.

Thank you so much for your help @ludeknovy, I really appreciate it!

**Reply by ludeknovy** [reply]
Glad we resolved it. How come that the `;` was used as delimiter?
I think I will add a note about the delimiter into documentation.

Anyway if you have any feedback or suggestions, please share it with me here on github or contact me at ludek@jtlreporter.site

**Reply by jmodeorain14** [reply]
I was switching the delimiter in the jmeter.properties file because I was doing some post-processing of the results which used `;`.

I was able to revert to the standard `,` delimiter and update my post-processing steps accordingly so it still works.

Adding a note to the docs about supported delimiters would be useful.
I don't think many people will change the default delimiter, but you never know.

**Reply by ludeknovy** [reply]
https://github.com/ludeknovy/jtl-reporter-docs/pull/41
